<a href="https://colab.research.google.com/github/pingao2019/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/PINgAO_of_DS_Unit_1_Sprint_Challenge_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 2

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 is a variable that indicates whether or not a person's request for credit has been approved or denied. This is a good candidate for a y variable since we might want to use the other features to predict this one. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [68]:
import pandas as pd
coulumn_names=['A1','A2','A3','A4','A5','A6','A7','A8','A9',
               'A10','A11','A12','A13',
               'A14','A15','A16']

df =pd.read_csv('crx1.txt', header=None, names=coulumn_names)
print(df.shape)
df.head(2)

(690, 16)


A1     A2    A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.00  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.46  u  g  q  h  3.04  t   t    6   f   g  00043  560   +

In [3]:
 import numpy as np
 df =df.replace({'?': np.NaN})
 df.head(2)

A1     A2    A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.00  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.46  u  g  q  h  3.04  t   t    6   f   g  00043  560   +

In [0]:
#checking missing data
df.isnull().sum()

In [16]:
df.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

In [17]:
df['A2'] = df['A2'].astype(float)
df['A14'] = df['A14'].astype(float)
df.dtypes

A1      object
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14    float64
A15      int64
A16     object
dtype: object

In [30]:
pos= df[df['A16']=='+']
print(pos.shape)
pos.head()

(307, 16)


A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

In [31]:
neg= df[df['A16']=='-']
print(neg.shape)
neg.head(3)

(383, 16)


A1     A2   A3 A4 A5  A6  A7      A8 A9 A10  A11 A12 A13    A14  A15 A16
70  b  32.33  7.5  u  g   e  bb   1.585  t   f    0   t   s  420.0    0   -
71  b  34.83  4.0  u  g   d  bb  12.500  t   f    0   t   g    NaN    0   -
72  a  38.58  5.0  u  g  cc   v  13.500  t   f    0   t   g  980.0    0   -

In [0]:
import numpy as np


from scipy import stats
from scipy.stats import ttest_1samp, ttest_ind,chi2

In [0]:
confidence_intervals = []

for col in pos:
  # print(confidence_interval(mpg[col]))
  confidence_intervals.append(confidence_interval(pos[col]))

# Add the list of tuples to a dataframe
confidence_interval_df = pd.DataFrame(confidence_intervals)
# rename the dataframe columns
confidence_interval_df.columns = ['lower_bound', 'mean', 'upper_bound']

confidence_interval_df.head()

In [0]:
def confidence_interval(data, confidence_level=.95):

  sample = np.array(data)
  
  s = np.std(sample, ddof=1)
  
  n = len(sample)
 
  x_bar = np.mean(sample)
  
  t = stats.t.ppf((1+confidence_level)/2, n-1)

  standard_error = s / np.sqrt(n)
  margin_of_error = t*standard_error

  upper_bound = x_bar + margin_of_error
  lower_bound = x_bar - margin_of_error

  return (lower_bound, x_bar, upper_bound)

In [51]:
CI = confidence_interval(pos['A3'])
CI

(5.290474713739196, 5.9049511400651475, 6.519427566391099)

In [0]:
confidence_interval_df['margin_of_error'] = confidence_interval_df['upper_bound'] - confidence_interval_df['mean']

confidence_interval_df.head()

In [0]:
# error bars

# x position of my error bars
x = range(0,2)
# Means
y = confidence_interval_df['mean']
# Margins of error
yerr = confidence_interval_df['margin_of_error']

plt.style.use('fivethirtyeight')
plt.errorbar(x, y, yerr, fmt='o')
plt.xticks(range(0,2), labels=["A2pos", "A2neg"])
 

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [60]:
pos.describe(include='number')

A2          A3  ...         A14            A15
count  305.000000  307.000000  ...  301.000000     307.000000
mean    33.720492    5.904951  ...  164.421927    2038.859935
std     12.809641    5.471485  ...  161.770675    7659.763941
min     13.750000    0.000000  ...    0.000000       0.000000
25%     23.170000    1.500000  ...    0.000000       0.000000
50%     30.500000    4.460000  ...  120.000000     221.000000
75%     41.330000    9.520000  ...  280.000000    1209.000000
max     76.750000   28.000000  ...  840.000000  100000.000000

[8 rows x 6 columns]

In [62]:
neg.describe(include='number')

A2          A3          A8         A11          A14          A15
count  373.000000  383.000000  383.000000  383.000000   376.000000   383.000000
mean    29.808231    3.839948    1.257924    0.631854   199.699468   198.605744
std     10.919291    4.337662    2.120481    1.900049   181.564835   671.608839
min     15.170000    0.000000    0.000000    0.000000     0.000000     0.000000
25%     22.000000    0.835000    0.125000    0.000000   100.000000     0.000000
50%     27.330000    2.210000    0.415000    0.000000   167.500000     1.000000
75%     34.830000    5.000000    1.500000    0.000000   272.000000    67.000000
max     80.250000   26.335000   13.875000   20.000000  2000.000000  5552.000000

In [0]:
from scipy import stats
from scipy.stats import ttest_1samp, ttest_ind,ch

In [20]:
ttest_ind(pos_A16['A2'], neg_A16['A2'],nan_policy='omit')

Ttest_indResult(statistic=4.2922156166315535, pvalue=2.0276637071781407e-05)

In [9]:
ttest_ind(pos_A16['A3'], neg_A16['A3'],nan_policy='omit')

Ttest_indResult(statistic=5.52998337614816, pvalue=4.551680702308068e-08)

In [11]:
ttest_ind(pos_A16['A8'], neg_A16['A8'],nan_policy='omit')

Ttest_indResult(statistic=8.935819983773698, pvalue=3.6710537401601785e-18)

In [12]:
ttest_ind(pos_A16['A11'], neg_A16['A11'],nan_policy='omit')

Ttest_indResult(statistic=11.667004222431277, pvalue=7.957718568079967e-29)

In [54]:
ttest_ind(pos_A16['A14'], neg_A16['A14'],nan_policy='omit')

Ttest_indResult(statistic=-2.6358251986645476, pvalue=0.008586135473979569)

In [55]:
ttest_ind(pos_A16['A15'], neg_A16['A15'],nan_policy='omit')

Ttest_indResult(statistic=4.680216020964486, pvalue=3.4520256956287944e-06)

Null Hypothesis: The average A2, A3,A8,A11,A14,A15 in positive group and in negative group are the same.
Alternative Hypothesis: the average A2, A3,A8,A11,A14,A15 in positive group and in negative group are different.
Confidence Level: 95%

from p-value all less than 0.05,  we can get conclusion, we can reject the Null Hypothesis that them in positive and negative group are same.
we accept the  Alternative Hypothesis, saying they in two group are different.

But for A14, p-value=0.008586135473979569 and with smallest statistic(-2,63), the  biggest one,  it is less different from two groups.
For A11(A eleven), giving smallest p-value and biggest statistic(11.67), suggest it has biggest difference in two groups. 

In [0]:
# chi2 for category analysis

In [58]:
df.describe(exclude='number')

A1   A4   A5   A6   A7   A9  A10  A12  A13  A16
count   678  684  684  681  681  690  690  690  690  690
unique    2    3    3   14    9    2    2    2    3    2
top       b    u    g    c    v    t    f    f    g    -
freq    468  519  519  137  399  361  395  374  625  383

In [0]:
def process_hours(df):
  cut_points = ['t','f']
  label_names = ["t","f"]
  df['A9_new'] = pd.cut(df['A9'], cut_points,labels=label_names)
  return df


In [79]:
data = process_hours(df)
A9_by_A16 = data[['A16', 'A9_new']]
A9_by_A16.head()

ValueError: ignored

In [0]:
A9_by_A16['A16'].value_counts()

In [0]:
contingency_table = pd.crosstab(A9_by_A16['A16'],
    A9_by_A16['A9_new'],
    margins = True)

contingency_table

In [0]:
poscount = contingency_table.iloc[0][0:2].values
negcount = contingency_table.iloc[1][0:2].values

#Plots the bar chart
fig = plt.figure(figsize=(10, 5))
sns.set(font_scale=1.8)
categories = ["t","f"]
p1 = plt.bar(categories, malecount, 0.55, color='#d62728')
p2 = plt.bar(categories, femalecount, 0.55, bottom=malecount)
plt.legend((p2[0], p1[0]), ('pos', 'neg'))
plt.xlabel('A9')
plt.ylabel('Count')
plt.show()

In [0]:
col_sums = contingency_table.iloc[2,0:2].values


In [0]:
total = contingency_table.loc['All', 'All']
print("Total number of observations:", total)

expected = []
for i in range(len(row_sums)):
  expected_row = []
  for column in col_sums:
    expected_val = column*row_sums[i]/total
    expected_row.append(expected_val)
  expected.append(expected_row)

In [0]:
contingency = pd.crosstab(9_by_A16['A16'], 
                          9_by_A16['A9_new'])
contingency = contingency.values

In [0]:
chi_squared = ((contingency - expected)**2/(expected)).sum()
print(f"Chi-Squared: {chi_squared}")

In [0]:
p_value = stats.chi2.sf(chi_squared, dof)
print(f"P-value: {p_value}")

In [0]:
chi_squared, p_value, dof, expected = stats.chi2_contingency(contingency)
print(f"Chi-Squared: {chi_squared}")
print(f"P-value: {p_value}")
print(f"Degrees of Freedom: {dof}") 
print("Expected: \n", np.array(expected)

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

from 2sample t test: 
Null Hypothesis: The average A2, A3,A8,A11,A14,A15 in positive group and in negative group are the same.
Alternative Hypothesis: the average A2, A3,A8,A11,A14,A15 in positive group and in negative group are different.
Confidence Level: 95%

from p-value all less than 0.05,  we can get conclusion, we can reject the Null Hypothesis that them in positive and negative group are same.
we accept the  Alternative Hypothesis, saying they in two group are different.

But for A14, p-value=0.008586135473979569 and with smallest statistic(-2,63), the  biggest one,  it is less different from two groups.
For A11(A eleven), giving smallest p-value and biggest statistic(11.67), suggest it has biggest difference in two groups.*

For chi2 test, even I know the process how to do chi2 test. I fail on  step 75th, so can not succed on get p_value. 

# Stretch Goals: 
Do these to get a 3. These are not required in order to pass the Sprint Challenge.

## Part 1: 

Make sure that all of your dataframe columns have the appropriate data types. *Hint:* If a column has the datatype of "object" even though it's made up of float or integer values, you can coerce it to act as a numeric column by using the `pd.to_numeric()` function. In order to get a 3 on this section make sure that your data exploration is particularly well commented, easy to follow, and thorough.

## Part 2:

Write functions that can calculate t-tests and chi^2 tests on all of the appropriate column combinations from the dataset. (Remember that certain tests require certain variable types.)

## Part 3: 

Calculate and report confidence intervals on your most important mean estimates (choose at least two). Make some kind of a graphic or visualization to help us see visually how precise these estimates are.

In [0]:
# You can work the stretch goals down here or back up in their regular sections
# just make sure that they are labeled so that we can easily differentiate
# your main work from the stretch goals.